### PR TITLE
Cleaner Saving

### PR DESCRIPTION
**Pull request by danyoungday**
Changed the save interface for prescriptors and changed how results are logged in the csv.

Now we store prescriptors and load prescriptors using a single file. We should be able to extract candidate id from this file as well to pair with the prescriptors.

Results are logged in one giant csv now with a generations column rather than 1 csv per generation.

### CODE REVIEW

File: examples/bnh/demo.ipynb [hunk 1]
```diff
@@ -170,12 +169,14 @@
     "    x2_col = []\n",
     "    gen_col = []\n",
     "    seen = set()\n",
+    "    results_df = pd.read_csv(results_dir / \"results.csv\")\n",
+    "    pop_dict = prescriptor_factory.load_population(results_dir / \"population\")\n",
     "    for gen in gens:\n",
-    "        df = pd.read_csv(results_dir / f\"{gen}.csv\")\n",
+    "        df = results_df[results_df[\"gen\"] == gen]\n",
```
Comment (danyoungday) [inline examples/bnh/demo.ipynb:175]: We have to make some minor tweaks to our experimentation code to work now, but overall the file structure is MUCH cleaner! We're willing to allow a little bit more work in the experiments to not have thousands of files per experiment.

File: presp/evolution.py [hunk 2]
```diff
@@ -42,40 +42,65 @@ def __init__(self,
         :param save_path: The path to save the results to.
         :param prescriptor_factory: The factory to create prescriptors with.
         :param evaluator: The evaluator used to evaluate the prescriptors.
-        :param seed_dir: Optional path to a directory containing seed candidates.
+        :param seed_path: Optional path to a file holding seed candidates.
```
Comment (danyoungday) [inline presp/evolution.py:45]: Seeds come from a single file now rather than a directory

File: presp/evolution.py [hunk 3]
```diff
@@ -42,40 +42,65 @@ def __init__(self,
...
+                                           "cv",
+                                           "rank",
+                                           "distance",
+                                           *self.evaluator.outcomes])
+        results_df.to_csv(self.save_path / "results.csv", index=False)
+        if self.validator:
+            val_df = pd.DataFrame(columns=["gen", "cand_id", *self.validator.outcomes])
+            val_df.to_csv(self.save_path / "validation.csv", index=False)
+        self.save_cache = []
+
```
Comment (danyoungday) [inline presp/evolution.py:85]: Here we set up our new `results_df` to be appended to during our logging. We also have a save cache of the candidates we want to save to file.

File: presp/evolution.py [hunk 4]
```diff
@@ -42,40 +42,65 @@ def __init__(self,
...
-            for seed_file in self.seed_dir.glob("*"):
-                candidate = self.prescriptor_factory.load(seed_file)
-                candidate.cand_id = seed_file.stem
+        if self.seed_path is not None:
+            print(f"Loading seeds from {self.seed_path}")
+            seed_dict = self.prescriptor_factory.load_population(self.seed_path)
+
+            def sort_cand_id(cand_id):
+                return int(cand_id.split("_")[0]), int(cand_id.split("_")[1])
+            sorted_ids = sorted(seed_dict.keys(), key=sort_cand_id)
```
Comment (danyoungday) [inline presp/evolution.py:97]: We sort by id here just for convenience

File: presp/evolution.py [hunk 5]
```diff
@@ -42,40 +42,65 @@ def __init__(self,
...
+
+            # We don't want duplicates cand_ids messing things up in our population
+            # We make the seeds generation 0 to show they are seeds
+            for i, cand_id in enumerate(sorted_ids):
+                candidate = seed_dict[cand_id]
+                candidate.cand_id = "0_" + str(i)
```
Comment (danyoungday) [inline presp/evolution.py:103]: We rename the seeds to `0_i` here so that our seed names can't cause non-unique candidate ids

File: presp/evolution.py [hunk 6]
```diff
@@ -169,26 +197,26 @@ def record_results(self):
                 row[outcome] = metric
             rows.append(row)
 
-        df = pd.DataFrame(rows)
-        df.to_csv(self.save_path / f"{self.generation}.csv", index=False)
+        results_df = pd.DataFrame(rows)
+        results_df.to_csv(self.save_path / "results.csv", mode="a", index=False, header=False)
+
+        self.prescriptor_factory.save_population(self.save_cache, self.save_path / "population")
```
Comment (danyoungday) [inline presp/evolution.py:203]: Every generation we save the whole population. This is a bit ugly but we can fix this later, for now it causes negligible slow downs.

File: presp/prescriptor/base.py [hunk 7]
```diff
@@ -58,16 +58,16 @@ def mutation(self, candidate: Prescriptor, mutation_rate: float, mutation_factor
...
     @abstractmethod
-    def save(self, candidate: Prescriptor, path: Path):
+    def save_population(self, population: list[Prescriptor], path: Path):
         """
-        Save the prescriptor to file.
-        :param prescriptor: The prescriptor to save.
-        :param path: The path to save the prescriptor to.
+        Saves a population of prescriptors to file.
+        :param population: The population of prescriptors to save.
+        :param path: The path to save the prescriptors to.
```
Comment (danyoungday) [inline presp/prescriptor/base.py:65]: function has to change to take in a population rather than a single candidate

File: presp/prescriptor/direct.py [hunk 8]
```diff
@@ -74,15 +75,23 @@ def mutation(self, candidate: DirectPrescriptor, mutation_rate: float, mutation_
         # Is this necessary?
         candidate.genome = genome
 
-    def save(self, candidate: DirectPrescriptor, path: Path):
+    def save_population(self, population: list[DirectPrescriptor], path: Path):
+        pop_dict = {cand.cand_id: cand.genome for cand in population}
         with open(path, "wb") as f:
             # Save the genome as a tensor
-            np.save(f, candidate.genome)
+            np.savez(f, **pop_dict)
```
Comment (danyoungday) [inline presp/prescriptor/direct.py:82]: We use numpy's `savez` function to do all the heavy lifting for us

File: presp/prescriptor/nn.py [hunk 9]
```diff
@@ -131,16 +131,22 @@ def mutation(self, candidate: NNPrescriptor, mutation_rate: float, mutation_fact
...
 
-    def save(self, candidate: NNPrescriptor, path: Path):
+    def save_population(self, population: list[NNPrescriptor], path: Path):
         """
-        Saves the torch model to file.
+        Put population into dict of cand_id -> NNPrescriptor and then save as a torch file.
         """
-        torch.save(candidate.model.state_dict(), path)
+        pop_dict = {cand.cand_id: cand.model.state_dict() for cand in population}
+        torch.save(pop_dict, path)
```
Comment (danyoungday) [inline presp/prescriptor/nn.py:139]: Easy peasy in pytorch

File: presp/evolution.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (danyoungday) [reply presp/evolution.py:203]: We want to save them all each time for now. This will need to be fixed later.

File: tests/test_evolution.py [hunk 10]
```diff
@@ -21,14 +22,12 @@ class TestInitialPopulation(unittest.TestCase):
...
-        if Path("tests/seeds").exists():
-            shutil.rmtree(Path("tests/seeds"))
+        Path("tests/seeds").unlink(missing_ok=True)
 
     def tearDown(self):
         if Path("tests/temp").exists():
             shutil.rmtree(Path("tests/temp"))
-        if Path("tests/seeds").exists():
-            shutil.rmtree(Path("tests/seeds"))
+        Path("tests/seeds").unlink(missing_ok=True)
```
[the remark this answers is not in the document]
Comment (danyoungday) [reply tests/test_evolution.py:30]: We know `tests/seeds` will be a file, but for the future we may want to start using file extensions for our saving/loading
